Allow carryall pickup orders on deployed vehicles

pchote opened the pull request:
Fixes #17211.
Fixes #17209.
Fixes #17248.

~Depends on #17214 and provides a testcase for https://github.com/OpenRA/OpenRA/pull/17214#discussion_r334230615.~

pchote committed:
Add UndeployOnPickup to GrantConditionOnDeploy.

pchote committed:
Allow carryall pickup orders on deployed vehicles.

tovl commented on OpenRA.Mods.Common/Activities/PickupUnit.cs:
```diff
@@ -74,26 +76,31 @@ public override bool Tick(Actor self)
 				return true;
 			}
 
+			// Abort if the cargo has canceled the reservation
```
It seems this entire branch can be removed, because the `cargo != carryall.Carryable` in L62 should already catch all the cases that this branch would catch. When the carryable is unreserved, cargo is set to null.

pchote replied:
Done.

pchote commented:
Updated, rebased, and swapping with #17214 on the milestone.

matjaeck commented:
This introduces a minor glitch when ordering to pick up a unit while already carrying one:

<details>

![carryglitch](https://user-images.githubusercontent.com/37185497/67079794-17e75080-f194-11e9-91fd-fc81b83a2947.gif)

</details>

Another thing I noticed is that units will undeploy for pickup even when they are disabled by EMP.

The fix for #17209 seems to work.

pchote commented:
> Another side effect is that infantry units do not "stick to a subcell" anymore while moving. This seems however also to be the case in the original:

I may be missing something in the animations, but the "PR" behaviour looks like the standard OpenRA release behaviour to me. Bleed behaves the same for me, I can't reproduce the behaviour in the "bleed" animation.

matjaeck commented:
> I may be missing something in the animations, but the "PR" behaviour looks like the standard OpenRA release behaviour to me.

You're right - sorry for the confusion. I don't know what causes it that they move in this fixed formation but I can reproduce it on ae34410c8077a45d3dae20419aa63659cf0718a6  after moving the 5-stack of infantry around a bit. But it's not related to this PR.

tovl commented on OpenRA.Mods.Common/Activities/PickupUnit.cs:
```diff
@@ -43,16 +42,19 @@ public PickupUnit(Actor self, Actor cargo, int delay)
 
 		protected override void OnFirstRun(Actor self)
 		{
 			carryall.ReserveCarryable(self, cargo);
```
Replacing the contents of `OnFirstRun` with:
```
			if (carryall.ReserveCarryable(self, cargo))
			{
				// Fly to the target and wait for it to be locked for pickup
				// These activities will be cancelled and replaced by Land once the target has been locked
				QueueChild(new Fly(self, Target.FromActor(cargo)));
				QueueChild(new FlyIdle(self, tickIdle: false));
			}
```
resolves the glitch when already carrying another unit

tovl commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnDeploy.cs:
```diff
@@ -157,6 +160,17 @@ Activity IWrapMove.WrapMove(Activity moveInner)
+			if (!Info.UndeployOnPickup || deployState == DeployState.Undeployed)
+				return true;
+
+			if (deployState == DeployState.Deployed)
```
Adding `&& !IsTraitPaused && !IsTraitDisabled` here will make sure that EMPed units won't unemploy.

pchote replied:
Disabling a trait is supposed to behave equivalently to the trait not being defined at all, so the disabled check belongs on the `return true` case above.

tovl replied:
That does hint at a bit of a conceptual flaw though. If the unit is unable to be carried in its current deploy state, then losing the ability to transition between states doesn't make it suddenly okay to be carried while deployed. The correct behaviour when semi-permanently losing the ability to undeploy when `UndeployOnPickup == true` would be to refuse the pick-up altogether, i.e. we should return `LockResponse.Failed` in that case.

pchote replied:
The disabled filtering is actually already handled in the `Carryall.LockForPickup` so it doesn't matter what we are doing here in any case.

pchote committed:
Disable Carryable while submerged.

pchote commented:
Updated:
* Fixed `PickupUnit.OnFirstRun` as suggested by @tovl for correctness
* Disabled the pickup cursor for carryalls with cargo
* Fixed pickup of EMPed deployed units
* Fixed pickup of submerged units

matjaeck commented:
> Disabled the pickup cursor for carryalls with cargo

I think this is too strict. The old behavior allowed you to queue multiple transport runs but now pickup orders become impossible to queue. Otherwise the fixes LGTM.

pchote committed:
Prevent movement pausing at invalid position.

pchote commented:
I hadn't considered that, good point. Reverted.

tovl reviewed (approved):
:+1:

pchote commented:
Filed #17254 to merge into prep, so bumping to Next + 1.
